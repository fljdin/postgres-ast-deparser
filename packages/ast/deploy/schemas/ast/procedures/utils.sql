-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
