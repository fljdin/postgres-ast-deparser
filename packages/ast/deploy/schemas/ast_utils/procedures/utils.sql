-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
