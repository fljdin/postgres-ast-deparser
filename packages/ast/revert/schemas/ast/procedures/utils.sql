-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
