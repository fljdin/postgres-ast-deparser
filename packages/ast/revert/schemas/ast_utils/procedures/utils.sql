-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
