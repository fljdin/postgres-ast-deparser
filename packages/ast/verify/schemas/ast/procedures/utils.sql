-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
