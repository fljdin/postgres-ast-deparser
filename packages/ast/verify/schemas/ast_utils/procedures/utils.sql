-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
