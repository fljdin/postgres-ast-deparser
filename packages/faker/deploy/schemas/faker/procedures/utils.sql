-- Deploy schemas/faker/procedures/utils to pg

-- requires: schemas/faker/schema
-- requires: schemas/faker/tables/dictionary/table

BEGIN;


CREATE FUNCTION faker.words() returns text as $$
SELECT 

(SELECT word FROM faker.dictionary 
WHERE type = 'adjectives'
OFFSET floor( random() * (select count(*) from faker.dictionary WHERE type = 'adjectives' ) ) LIMIT 1)
|| ' ' ||
(SELECT word FROM faker.dictionary 
WHERE type = 'colors'
OFFSET floor( random() * (select count(*) from faker.dictionary WHERE type = 'colors' ) ) LIMIT 1)
|| ' ' ||
(SELECT word FROM faker.dictionary 
WHERE type = 'animals'
OFFSET floor( random() * (select count(*) from faker.dictionary WHERE type = 'animals' ) ) LIMIT 1)
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.word_type() returns text as $$
SELECT (CASE (RANDOM() * 2)::INT
      WHEN 0 THEN 'adjectives'
      WHEN 1 THEN 'colors'
      WHEN 2 THEN 'animals'
    END);
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.word(wordtype text default null) returns text as $$
DECLARE
  vword text;
  vtype text;
BEGIN
  IF (wordtype IS NOT NULL) THEN 
    vtype = wordtype;
  ELSE
    vtype = faker.word_type();
  END IF;

SELECT word FROM faker.dictionary 
WHERE type = vtype
OFFSET floor( random() * (select count(*) from faker.dictionary WHERE type = vtype ) ) LIMIT 1
INTO vword;

RETURN vword;

END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.sentence(unit text default 'word', min int default 7, max int default 20, cat text default 'lorem', period text default '.') returns text as $$
DECLARE
  num int = faker.integer(min, max);
  txt text;
  vtype text;
  n int;
  c int;
BEGIN

  IF (unit = 'word') THEN
    txt = initcap(faker.word(cat));
    FOR n IN 
    SELECT * FROM generate_series(1, num) g(n)
    LOOP 
      txt = txt || ' ' || faker.word(cat);
    END LOOP;
    RETURN txt || period;
  ELSEIF (unit = 'char' OR unit = 'chars') THEN 
    txt = initcap(faker.word(cat));
    c = char_length(txt);
    IF (c = num) THEN 
      RETURN concat(txt, period);
    END IF;
    IF (c > num) THEN 
      RETURN substring(txt from 1 for num) || period;
    END IF;
    WHILE (c < num)
    LOOP 
      txt = txt || ' ' || faker.word(cat);
      c = char_length(txt);
    END LOOP;
    IF (c = num) THEN 
      RETURN txt || period;
    END IF;
    IF (c > num) THEN 
      RETURN substring(txt from 1 for num) || period;
    END IF;
  END IF;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.paragraph(unit text default 'word', min int default 7, max int default 20, cat text default 'lorem') returns text as $$
DECLARE
  num int = faker.integer(min, max);
  txt text;
  words text[];
  n int;
  needscaps boolean = false;
BEGIN
  txt = faker.sentence(unit, min, max, cat, '');
  words = string_to_array(txt, ' ');
  txt = '';

  FOR n IN
  SELECT * FROM generate_series(1, cardinality(words)) g(n)
  LOOP 
      IF (needscaps IS TRUE) THEN
        txt = concat(txt, ' ', initcap(words[n]));
      ELSE
        txt = concat(txt, ' ', words[n]);
      END IF;

      IF (faker.integer(1,100) > 70) THEN 
        txt = txt || '.';
        needscaps = true;
      ELSE
        needscaps = false;
      END IF;

  END LOOP;

  IF (trim(txt) ~ '\.$') THEN 
    RETURN trim(txt);
  ELSE
    RETURN trim(txt || '.');
  END IF;

END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.email() returns text as $$
SELECT
  faker.word() || (RANDOM() * 100)::INT || '@' || (
    CASE (RANDOM() * 3)::INT
      WHEN 0 THEN 'gmail'
      WHEN 1 THEN 'hotmail'
      WHEN 2 THEN 'yahoo'
      WHEN 3 THEN faker.word()
    END
  ) || '.com' AS email
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.uuid() returns uuid as $$
SELECT
  uuid_generate_v4();
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.token(bytes int default 16) returns text as $$
SELECT
  encode( gen_random_bytes( bytes ), 'hex' )
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.password() returns text as $$
DECLARE
  chars text[] =  regexp_split_to_array('!@#$%^&*():";''<>?,./~`'::text, '');
  num_special int = faker.integer(0, 4);
  n int = 0;
  pass text;
BEGIN

  IF (num_special = 0) THEN
    pass = encode( gen_random_bytes( 16 ), 'hex' );
  ELSE
    pass = encode( gen_random_bytes( 4 ), 'hex' );
    FOR n IN
    SELECT * FROM generate_series(1, num_special) g(n)
    LOOP
      pass = pass ||
      encode( gen_random_bytes( faker.integer(1,4) ), 'hex' ) ||
      chars[ faker.integer(1, cardinality(chars)) ] ||
      encode( gen_random_bytes( faker.integer(1,4) ), 'hex' );

    END LOOP;
  END IF;


  RETURN pass;

END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.hostname() returns text as $$
SELECT
  faker.word() || '.' || (
    CASE (RANDOM() * 4)::INT
      WHEN 0 THEN 'com'
      WHEN 1 THEN 'net'
      WHEN 2 THEN 'io'
      WHEN 3 THEN 'org'
      WHEN 4 THEN 'co'
    END
  )
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.time_unit() returns text as $$
SELECT
  (
    CASE (RANDOM() * 5)::INT
      WHEN 0 THEN 'millisecond'
      WHEN 1 THEN 'second'
      WHEN 2 THEN 'minute'
      WHEN 3 THEN 'hour'
      WHEN 4 THEN 'day'
      WHEN 5 THEN 'week'
    END
  )
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.float(min int default 0, max int default 100) returns float as $$
DECLARE
  num float;
BEGIN
  min = ceil(min);
  max = floor(max);
  num = (RANDOM() * ( max - min + 1 )) + min;
  RETURN num;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.integer(min int default 0, max int default 100) returns int as $$
DECLARE
  num int;
BEGIN
  min = ceil(min);
  max = floor(max);
  num = floor(RANDOM() * ( max - min + 1 )) + min;
  RETURN num;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.date(future boolean default false) returns date as $$
DECLARE
  d date;
BEGIN
IF (future) THEN 
  d = now()::date + 10;
ELSE
  d = now()::date - 10;
END IF;
RETURN d;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.interval() returns interval as $$
DECLARE
  ival text;
BEGIN
  SELECT faker.time_unit() INTO ival;
  ival = (RANDOM() * 100)::text || ' ' || ival;
  RETURN ival::interval;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.timestamptz(future boolean default false) returns timestamptz as $$
DECLARE
  ival interval;
  t timestamptz;
BEGIN
  ival = faker.interval();
  IF (future) THEN
    SELECT now() + ival INTO t;
  ELSE
    SELECT now() - ival INTO t;
  END IF;
  RETURN t;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.mime() returns text as $$
  SELECT faker.word('mime');
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.ext(mime text default faker.mime()) returns text as $$
DECLARE 
  ext text;
BEGIN
  IF (mime IS NULL) THEN 
    ext = faker.word(faker.mime());
  ELSE
    ext = faker.word(mime);
  END IF;
  RETURN ext;
END;
$$
LANGUAGE 'plpgsql';


CREATE FUNCTION faker.image_mime() returns text as $$
SELECT
  (
    CASE (RANDOM() * 3)::INT
      WHEN 0 THEN 'image/svg+xml'
      WHEN 1 THEN 'image/png'
      WHEN 2 THEN 'image/gif'
      WHEN 3 THEN 'image/jpeg'
    END
  )
$$
LANGUAGE 'sql';

CREATE FUNCTION faker.image(width int default null, height int default null) returns image as $$
DECLARE
  w int;
  h int;
  obj jsonb = '{}'::jsonb;
  url text;
BEGIN
  IF (width IS NULL) THEN 
    w = faker.integer(800,1200);
  ELSE
    w = width;
  END IF;
  IF (height IS NULL) THEN 
    h = faker.integer(800,1200);
  ELSE
    h = height;
  END IF;

  url = concat('https://picsum.photos/', w::text, '/', h::text);

  obj = jsonb_set(obj, '{url}', to_jsonb(url));
  obj = jsonb_set(obj, '{mime}', to_jsonb(faker.image_mime()));

  RETURN obj;

END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.url(mime text default null) returns url as $$
DECLARE
  obj jsonb = '{}'::jsonb;
  url text;
BEGIN
  url = concat('https://', faker.hostname(), '/', faker.word(), '.', faker.ext(mime));
  RETURN url;
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.upload(mime text default null) returns upload as $$
BEGIN
  RETURN faker.url(mime);
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.ip(mime text default null) returns text as $$
BEGIN
  RETURN 
    array_to_string(ARRAY[
      faker.integer(0,255),
      faker.integer(0,255),
      faker.integer(0,255),
      faker.integer(0,255)
    ]::text[], '.');
END;
$$
LANGUAGE 'plpgsql';

CREATE FUNCTION faker.attachment(mime text default null) returns attachment as $$
DECLARE
  obj jsonb = '{}'::jsonb;
BEGIN
  IF (mime IS NULL) THEN
    mime = faker.mime();
  END IF;
  obj = jsonb_set(obj, '{url}', to_jsonb(faker.url(mime)::text));
  obj = jsonb_set(obj, '{mime}', to_jsonb(mime));
  RETURN obj;
END;
$$
LANGUAGE 'plpgsql';

COMMIT;
