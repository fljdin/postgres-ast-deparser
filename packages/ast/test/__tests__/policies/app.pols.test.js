import { getConnections } from '../../utils';

let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {}
});

const getPolicyResult = async (name, vars) => {
  const [{ deparse }] = await db.any(
    `
  SELECT deparser.deparse(
    ast_helpers.create_policy_template(
      $1::text,
      $2::jsonb
    ))`,
    [name, JSON.stringify(vars)]
  );

  return deparse;
};

it('1 core database', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'mem_priv',
    acl_table: 'memberships_acl'
  });
  expect(result).toMatchSnapshot();
});

it('1 core table', async () => {
  const result = await getPolicyResult('acl_field_join', {
    entity_field: 'database_id',
    sel_obj: true,
    sel_field: 'id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    acl_join_field: 'entity_id',
    obj_schema: 'collections_public',
    obj_table: 'database',
    obj_field: 'owner_id'
  });
  expect(result).toMatchSnapshot();
});

//

it('groups READ', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'group_memberships_acl'
  });
  expect(result).toMatchSnapshot();
});

it('groups WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'group_memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('group_memberships WRITE', async () => {
  const result = await getPolicyResult('acl_field_join', {
    entity_field: 'entity_id',
    sel_obj: true,
    sel_field: 'entity_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    acl_join_field: 'actor_id',
    obj_schema: 'priv',
    obj_table: 'group_memberships_acl',
    obj_field: 'actor_id',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('group_memberships READ', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'entity_id',
    acl_schema: 'priv',
    acl_table: 'group_memberships_acl'
  });
  expect(result).toMatchSnapshot();
});

it('actions READ', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl'
  });
  expect(result).toMatchSnapshot();
});

it('actions WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('action_items WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('action_items INSERT', async () => {
  const result = await getPolicyResult('acl_field_join', {
    entity_field: 'action_id',
    sel_obj: true,
    sel_field: 'id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    acl_join_field: 'entity_id',
    obj_schema: 'public',
    obj_table: 'actions',
    obj_field: 'owner_id',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

/*
this one is a bit interesting, since it's cross referencing 
two different ACL tables, leveraging actor_id as the actual organization

essentially this reads as:
find entities where the current user is an admin (this would include their own account)
*/

it('app_memberships READ', async () => {
  const result = await getPolicyResult('acl_field_join', {
    entity_field: 'entity_id',
    sel_obj: true,
    sel_field: 'actor_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    acl_join_field: 'entity_id',
    obj_schema: 'priv',
    obj_table: 'app_memberships_acl',
    obj_field: 'actor_id',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('app_memberships ADMIN READ/WRITE', async () => {
  const result = await getPolicyResult('acl_exists', {
    acl_schema: 'priv',
    acl_table: 'app_memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('emails READ', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl'
  });
  expect(result).toMatchSnapshot();
});

it('emails WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'owner_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('memberships READ', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'entity_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl'
  });
  expect(result).toMatchSnapshot();
});

it('memberships WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'entity_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('org_profiles WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'organization_id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});

it('users WRITE', async () => {
  const result = await getPolicyResult('acl_field', {
    entity_field: 'id',
    acl_schema: 'priv',
    acl_table: 'memberships_acl',
    is_admin: true
  });
  expect(result).toMatchSnapshot();
});
