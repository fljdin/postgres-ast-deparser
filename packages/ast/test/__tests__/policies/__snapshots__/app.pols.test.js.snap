// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 core database 1`] = `"owner_id IN (SELECT acl.entity_id FROM mem_priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`1 core table 1`] = `"database_id IN (SELECT obj.id FROM priv.memberships_acl AS acl INNER JOIN collections_public.database AS obj ON acl.entity_id = obj.owner_id WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`action_items INSERT 1`] = `"action_id IN (SELECT obj.id FROM priv.memberships_acl AS acl INNER JOIN public.actions AS obj ON acl.entity_id = obj.owner_id WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`action_items WRITE 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`action_items verifier SELECT 1`] = `"action_id IN (SELECT obj.id FROM priv.memberships_acl AS acl INNER JOIN public.actions AS obj ON acl.entity_id = obj.owner_id WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`actions INSERT entity_has_permission 1`] = `"owner_id IN (SELECT acl.actor_id FROM priv.app_memberships_acl AS acl WHERE acl.actor_id = owner_id AND (acl.permissions & '1010101') = '1010101')"`;

exports[`actions READ 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`actions WRITE 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`app_memberships ADMIN READ/WRITE 1`] = `"EXISTS (SELECT 1 FROM priv.app_memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`app_memberships READ 1`] = `"entity_id IN (SELECT obj.actor_id FROM priv.memberships_acl AS acl INNER JOIN priv.app_memberships_acl AS obj ON acl.entity_id = obj.actor_id WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`emails READ 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`emails WRITE 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`group_memberships READ 1`] = `"entity_id IN (SELECT acl.entity_id FROM priv.group_memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`group_memberships WRITE 1`] = `"entity_id IN (SELECT obj.entity_id FROM priv.memberships_acl AS acl INNER JOIN priv.group_memberships_acl AS obj ON acl.actor_id = obj.actor_id WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`groups READ 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.group_memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`groups WRITE 1`] = `"owner_id IN (SELECT acl.entity_id FROM priv.group_memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`memberships READ 1`] = `"entity_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id())"`;

exports[`memberships WRITE 1`] = `"entity_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`org_profiles WRITE 1`] = `"organization_id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;

exports[`users WRITE 1`] = `"id IN (SELECT acl.entity_id FROM priv.memberships_acl AS acl WHERE acl.actor_id = jwt_public.current_user_id() AND acl.is_admin IS TRUE)"`;
